ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -125,33 +125,51 @@ let set_diagnostics rpc doc =
-                    | _ :: _ ->
-                      Some
-                        (List.map error.sub ~f:(fun (sub : Loc.msg) ->
-                             let location =
-                               let range = Range.of_loc sub.loc in
-                               Location.create ~range ~uri
-                             in
-                             let message = message Loc.print_sub_msg sub in
-                             DiagnosticRelatedInformation.create ~location
-                               ~message))
-                  in
-                  let message = message Loc.print_main error in
-                  create_diagnostic ?relatedInformation range message ~severity))
+              let merlin_diagnostics =
+                List.map errors ~f:(fun (error : Loc.error) ->
+                    let loc = Loc.loc_of_report error in
+                    let range = Range.of_loc loc in
+                    let severity =
+                      match error.source with
+                      | Warning -> DiagnosticSeverity.Warning
+                      | _ -> DiagnosticSeverity.Error
+                    in
+                    let message ppf m =
+                      String.trim (Format.asprintf "%a@." ppf m)
+                    in
+                    let relatedInformation =
+                      match error.sub with
+                      | [] -> None
+                      | _ :: _ ->
+                        Some
+                          (List.map error.sub ~f:(fun (sub : Loc.msg) ->
+                               let location =
+                                 let range = Range.of_loc sub.loc in
+                                 Location.create ~range ~uri
+                               in
+                               let message = message Loc.print_sub_msg sub in
+                               DiagnosticRelatedInformation.create ~location
+                                 ~message))
+                    in
+                    let message = message Loc.print_main error in
+                    create_diagnostic ?relatedInformation range message
+                      ~severity)
+              in
+              let holes_as_err_diags =
+                Query_commands.dispatch pipeline Holes
+                |> List.map ~f:(fun (loc, typ) ->
+                       let range = Range.of_loc loc in
+                       let severity = DiagnosticSeverity.Error in
+                       let message =
+                         "This typed hole should be replaced with an \
+                          expression of type " ^ typ
+                       in
+                       (* we set specific diagnostic code = "hole" to be able to
+                          filter through diagnostics easily *)
+                       create_diagnostic ~code:(`String "hole") range message
+                         ~severity)
+              in
+              List.append holes_as_err_diags merlin_diagnostics)
```

Comment:
We should make sure that all diagnostics are ordered by location.

A test case for this would be useful too.

Reply:
I'll add test cases for new diagnostics and diagnostics ordering. Out of curiosity: why do we want the diagnostics to be ordered?

Reply:
When one is walking through the document with "next diagnostic", we don't want the cursor to be jumping back and forth.